File: clojure/life.clj
(ns life
  (:use clojure.test))

(defn alive [] :alive)
(defn dead [] :dead)
(defn alive? [cell] (= (alive) cell))
(defn dead? [cell] (= (dead) cell))

(declare neighbours)

(defn next-gen [grid cell]
  (let [num-alive (count (filter alive? (neighbours grid)))]
    (cond
        (< num-alive 2) (dead)
        (= num-alive 2) (if (alive? cell) (alive) (dead))
        (= num-alive 3) (alive)
        :else (dead))))

(defn alive-neighbours [num]
  (concat (replicate num (alive)) (replicate (- 8 num) (dead))))

(deftest dies-with-no-neighbours
  (binding [neighbours (fn [_] (alive-neighbours 0))]
    (is (dead? (next-gen nil (alive))))))

(deftest succession
  (let [make-test (fn [[num-alive start expected]]
                    (binding [neighbours (fn [_] (alive-neighbours num-alive))]
                      (is (expected (next-gen nil start)))))]
    (doseq [spec [[0 (dead)  dead?]
                  [0 (alive) dead?]
                  [1 (dead)  dead?]
                  [1 (alive) dead?]
                  [2 (dead)  dead?]
                  [2 (alive) alive?]
                  [3 (dead)  alive?]
                  [3 (alive) alive?]
                  [4 (dead)  dead?]
                  [4 (alive) dead?]
                  [5 (dead)  dead?]
                  [5 (alive) dead?]
                  [8 (dead)  dead?]
                  [8 (alive) dead?]]]
      (make-test spec))))

(run-tests)

File: python/life.py
class Cell:
    def __init__(self, grid):
        self.grid = grid
    def is_dead(self):
        return not self.is_alive()
    def _num_living_neighbours(self):
        return len(filter(lambda n: n.is_alive(), self._neighbours()))
    def _neighbours(self):
        return self.grid.neighbours(self)

class Alive(Cell):
    def next_gen(self):
        if self._num_living_neighbours() == 2:
            return Alive(self.grid)
        return Dead(self.grid)
    def is_alive(self):
        return True

class Dead(Cell):
    def next_gen(self):
        return Dead(self.grid)
    def is_alive(self):
        return False

def with_neighbours(num):
    class StubGrid:
        def neighbours(self, cell):
            return [Alive(self) for n in range(0, num)] + [Dead(self) for n in range(0, 8-num)]
    return StubGrid()

def make_test(num_alive, start, expected):
    name = 'test_%s_with_%s_neighbours_%s' % (start.__name__, num_alive, expected)
    def test():
        next = start(with_neighbours(num_alive)).next_gen()
        assert getattr(next, expected)()
    globals()[name] = test

[make_test(*spec) for spec in [[0, Alive, 'is_dead'],
                               [0, Dead,  'is_dead'],
                               [1, Alive, 'is_dead'],
                               [1, Dead,  'is_dead'],
                               [2, Alive, 'is_alive'],
                               [2, Dead,  'is_dead'],
                               ]]
